cli/azd/pkg/project/framework_service_swa.go

```diff
@@ -0,0 +1,137 @@
+// leverages swa cli for building
+func NewSwaProject(
+	env *environment.Environment,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	swa swa.SwaCli,
+	framework FrameworkService,
+) CompositeFrameworkService {
+	return &swaProject{
+		env:           env,
+		console:       console,
+		commandRunner: commandRunner,
+		swa:           swa,
+		framework:     framework,
+	}
+}
+
+// NewSwaProjectAsFrameworkService is the same as NewSwaProject().(FrameworkService) and exists to support our
+// use of DI and ServiceLocators, where we sometimes need to resolve this type as a FrameworkService instance instead
+// of a CompositeFrameworkService as [NewSwaProject] does.
+func NewSwaProjectAsFrameworkService(
+	env *environment.Environment,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	swa swa.SwaCli,
+	framework FrameworkService,
+) FrameworkService {
+	return NewSwaProject(env, console, commandRunner, swa, framework)
+}
+
+func (p *swaProject) Requirements() FrameworkRequirements {
+	return FrameworkRequirements{
+		Package: FrameworkPackageRequirements{
+			RequireRestore: false,
+			RequireBuild:   true,
+		},
+	}
+}
+
+// Gets the required external tools for the project
+func (p *swaProject) RequiredExternalTools(context.Context) []tools.ExternalTool {
+	return []tools.ExternalTool{p.swa}
+}
+
+// Initializes the swa project
+func (p *swaProject) Initialize(ctx context.Context, serviceConfig *ServiceConfig) error {
+	return p.framework.Initialize(ctx, serviceConfig)
+}
+
+// Sets the inner framework service used for restore and build command
+func (p *swaProject) SetSource(inner FrameworkService) {
+	p.framework = inner
+}
+
+// Restores the dependencies for the swa project
+func (p *swaProject) Restore(
+	ctx context.Context,
+	serviceConfig *ServiceConfig,
+) *async.TaskWithProgress[*ServiceRestoreResult, ServiceProgress] {
+	// When the program runs the restore actions for the underlying project (containerapp),
+	// the dependencies are installed locally
+	return p.framework.Restore(ctx, serviceConfig)
+}
+
+// Builds the docker project based on the docker options specified within the Service configuration
+func (p *swaProject) Build(
+	ctx context.Context,
+	serviceConfig *ServiceConfig,
+	restoreOutput *ServiceRestoreResult,
+) *async.TaskWithProgress[*ServiceBuildResult, ServiceProgress] {
+	return async.RunTaskWithProgress(
+		func(task *async.TaskContextWithProgress[*ServiceBuildResult, ServiceProgress]) {
+
+			previewerWriter := p.console.ShowPreviewer(ctx,
+				&input.ShowPreviewerOptions{
+					Prefix:       "  ",
+					MaxLineCount: 8,
+					Title:        "Build SWA Project",
+				})
+			err := p.swa.Build(
+				ctx,
+				serviceConfig.Path(),
+				previewerWriter,
+			)
+			p.console.StopPreviewer(ctx, false)
+
+			if err != nil {
+				task.SetError(fmt.Errorf("swa build: %w", err))
+				return
+			}
+
+			task.SetResult(&ServiceBuildResult{
+				Restore: restoreOutput,
+			})
```

Comment:
Is there any other build output we can return here that could be helpful to the end users?

Reply:
Not that I know of.